feat(integrations): outlook

RobertoE91 opened the pull request:
This integration enables seamless email communication between your Botpress chatbot and Microsoft Outlook.

RobertoE91 committed:
feat(integrations): outlook

RobertoE91 commented on integrations/outlook/src/index.ts:
```diff
@@ -0,0 +1,252 @@
+        })
+
+        await client.createMessage({
+          tags: { 'outlook:id': notificationContent.id },
```
how can I access this tag? or how can I get the last message id for a given conversation?

franklevasseur replied:
have you tried listing messages and getting the last one ?

franklevasseur commented on integrations/outlook/src/index.ts:
```diff
@@ -0,0 +1,252 @@
+            },
+          })
+        },
+        image: async () => {
```
**issue:** please remove these content types from the definition if you don't intend to implement them

franklevasseur commented on integrations/outlook/src/index.ts:
```diff
@@ -0,0 +1,252 @@
+    // as plain text with a 200 response
+    // https://docs.microsoft.com/graph/webhooks#notification-endpoint-validation
+    if (req.query) {
+      const queryString = req.query as string
```
**chore:** there are node librairies to parse query string. This obfuscate the behavior of the function.

franklevasseur commented on integrations/outlook/src/index.ts:
```diff
@@ -0,0 +1,252 @@
+      return
+    }
+
+    body.value.forEach(async (message: any) => {
```
**issue:** please use the `for` operator

```ts
for (const message of body.value) { ... }
```
the will prevent floating promises rejections

also, I'd extract the content of the for loop in a dedicated function

franklevasseur replied:
**chore:** instead of doing manual data validation, please use `zod` to validate the request payload only once and then you can assume that the data has the desired format.

franklevasseur commented on integrations/outlook/src/client.ts:
```diff
@@ -0,0 +1,116 @@
+   * Subscribe webhook
+   */
+  public subscribeWebhook = async (webhookUrl: string, ctx: IntegrationContext): Promise<string> => {
+    const expirationDateTime = this.generateExpirationDate()
```
**question:** I don't get this. Will the integration stop working after 20 minutes ? Is it possible not to set an expiration date ?

franklevasseur replied:
Ok, I just saw your comment above. I answered it right under.

franklevasseur reviewed (commented):
Hey, sorry for the long delai!

I haven't tested the integration yet so I can't approve. I'll do it this week.

This is a first round of comments. I'll probably do a second pass after trying the integration.

I tried following the [conventional comments convention](https://conventionalcomments.org/).

Overall, I like your work. Good job !

Let me know if there's anything I can do to help you.

Frank

franklevasseur commented on integrations/outlook/src/client.ts:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  private generateExpirationDate = () => {
+    return moment.utc().add(20, 'minutes')
```
ok I think I get it. Basically, Outlook will call the webhook with an event to notify that the webhook is about to be expired ? Then you call the `handleLifecycleEvents` method ?

If so, is there any risk of race conditions ? Will we miss some messages in between webhook expiry ? Does outlook call **before** the webhook is expired or when it already is ?

If this is not safe enough, we can use Botpress recurring events to refresh the webhook every 5 min (keeping the expiry at 20 min)

anyway, let me know if I understand correctly and let's discuss this

RobertoE91 replied:
1- "Basically, Outlook will call the webhook with an event to notify that the webhook is about to be expired ? Then you call the handleLifecycleEvents method ?":
Yes, exactly. Outlook will send the lifecycle event **before** the webhook expires to prevent downtime.

2-"Will we miss some messages in between webhook expiry ?":
The webhook may continue for a while, giving you extra time to respond. However, eventually change notification delivery pauses, until you take the required action. Any notifications about resource changes that happen when the change notification delivery pauses and the time when the app successfully creates the subscription again would be lost.

3-"is there any risk of race conditions ?":
I think there is no risk of the race condition, Outlook sends a notification at a time and expects a successful response.
other than that, there was a situation in my local dev env where the "regiter" method of the integration was executed before unregistering the previous subscription and caused to receive duplicated messages.


extra notes: maximum time for the webhook to expire is  4,230 minutes (under 3 days)

franklevasseur replied:
I see no problem with keeping 20 minutes then! Let's keep it like that.

RobertoE91 replied:
ok, we're also implementing a way to check if there's an existing active subscripcion to avoid duplicated messages

allardy commented:
Please split your integration into different files like on slack, linear and github. It makes it easier to work and add features afterwards. A single file with everything is not practical

RobertoE91 committed:
resolve pr comments & files structure

allardy commented on integrations/outlook/src/definitions/index.ts:
```diff
@@ -0,0 +1,50 @@
+  schema: z.object({
+    tenantId: z.string({
+      description: "Azure AD tenant id",
+    }),
```
`z.string().describe()` is a bit clearer and makes the schema smaller

franklevasseur replied:
Yo @allardy, I try to use conventional comments when reviewing a PR. This is a **nitpick**

allardy commented on integrations/outlook/src/definitions/index.ts:
```diff
@@ -0,0 +1,50 @@
+      lastMessageId: z.string(),
+    }),
+  },
+} satisfies IntegrationDefinitionProps["states"];
```
Use `IntegrationProps from '.botpress'` to have your defined typings (otherwise you get generic ones)

franklevasseur replied:
same goes for all other files

allardy commented on integrations/outlook/src/misc/custom-schemas.ts:
```diff
@@ -0,0 +1,52 @@
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message: `Invalid email body: ${JSON.stringify(val, null, 3)}`,
+        path: ["body", "content"],
```
try to avoid using refine, we convert these zod schemas to json schemas and then back to zod and these validations are lost. Instead you can use `message: z.string().min(1, 'email message must be defined')` (if the original zod error message is not clear enough)

franklevasseur replied:
I agree here, this is a **chore**

allardy commented on integrations/outlook/src/misc/custom-schemas.ts:
```diff
@@ -0,0 +1,52 @@
+import * as z from "zod";
+
+export const notificationContentSchema = z
```
This schema feels a bit complex to use. May I suggest something a bit simpler? 
```js
const emailAddress = z.string().min(1, 'Email cannot be blank')
const emailOrNameEmail = z.union([
  emailAddress, 
  z.object({
    email: emailAddress,
    name: z.string().optional()
  })
]) // so you can just give an array of strings, or also an object with the name

const schema = z.object({
  from: emailOrNameEmail,
  to: z.array(emailOrNameEmail),
  cc: z.array(emailOrNameEmail),
  message: z.string().min(1)
})
```

It's easy to use a string for simplicity and also have an object for more complex setups

allardy commented on integrations/outlook/src/setup/handler.ts:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  return;
+};
```
can you add eslint & prettier in your vscode ? there's not supposed to be semicolon there

RobertoE91 committed:
Removed Not Implemented messages types

RobertoE91 committed:
Merge branch 'master' into re-integrations-outlook

franklevasseur committed:
chore(outlook): use pnpm + fix prettier and eslint

franklevasseur committed:
chore(outlook): cleanup typings

franklevasseur committed:
sync pnpm lock

RobertoE91 committed:
fix(general): pr comments

RobertoE91 committed:
feat(integration): Outlook actions send email n create calendar events

RobertoE91 committed:
additional features

- Added SVG Icon Outlook

- Changed package-json

-Added Logger and error handling send-email

-Added Logger and error handling create-event

bassamtantawi-botpress committed:
Title
